programs/survival/cam.sc

```diff
@@ -161,11 +169,107 @@ __survival_defaults(player) ->
          )
       )
    );
-   print(format('rb Cannot find a safe spot to land within 32 blocks.'));
+   display_title(player, 'actionbar', format('r Cannot find a safe spot to land within 32 blocks'));
    false;
 );
 
-// This fixes the case where the server has force-gamemode on, in which case the player would be put
-// into survival in the location they logged out in camera mode, bypassing the sign-off actions of the app.
-// __get_player_stored_takeoff_params evaluates true-y when player was in cam mode before logging out
-__on_player_connects(player) -> if( __get_player_stored_takeoff_params(player~'name'), modify(player, 'gamemode' , 'spectator' ));
+__restrict_flight(player) -> 
+(
+   if(global_max_flight_range > 0, 
+      config = __get_player_stored_takeoff_params(player~'name');
+      if(!config,  // gotta restrict it somehow if no config was saved
+         start_pos = player~'pos';
+         start_dim = player~'dimension',
+
+         start_pos = config:'pos'; 
+         start_dim = config:'dimension'
+      );
+      
+      __restrict_flight_tick(player, start_pos, start_dim)
+   )
+);
+
+__restrict_flight_tick(player, start_pos, start_dim) -> 
+(
+   if(player~'gamemode' != 'spectator', exit());
+
+   pp = player~'pos';
+   distance = __get_distance(copy(pp), copy(start_pos));
+   
+   if(distance>global_max_flight_range,
+      if(distance>global_max_flight_range*1.5,
+         new_pos = __get_restriction_new_pos(pp, start_pos);
+         modify( player, 'pos', new_pos); //in case they TP to another player
+         display_title(player, 'actionbar', format('y You are not allowed to go that far')),
+
+         new_motion = __get_restriction_new_motion(pp, start_pos, player~'motion');
+         modify( player, 'motion', new_motion),
+      )
+      
+   );
+
+   if(player~'dimension' != start_dim,
+      run('execute in minecraft:'+ start_dim +' run tp @s ~ ~ ~');
+      modify( player, 'pos', start_pos);
+      display_title(player, 'actionbar', format('y You are not allowed to change dimensions'));
+   );
+
+   schedule(1, '__restrict_flight_tick', player, start_pos, start_dim);
+);
+
+__get_distance(pos1, pos2)->
+(
+   // messure flat distance for square and cylindrical
+   if (global_restriction_shape!=2,
```

Comment:
Is this meant to be `!=` or `==` for cylindrical distance?

Reply:
Code is correct, if ==2 it is the spherical case, in which we should not collapse the 3d space on a 2d plane.

Reply:
This ^